File: 12/12.rs
use std::cmp::min;
use std::collections::{HashMap, HashSet, VecDeque};
use std::fs;
use std::time::Instant;

type Coord = u8;
type Height = u16;

fn main() {
    let start = Instant::now();

    let (map, s, strt, end) = parse_input("12/input.txt");

    let p1 = part1(&map, s, end);
    println!("part1: {}", p1);
    println!("part2: {}", part2(&map, &strt, end, p1));

    println!("time: {:?}", start.elapsed());
}

fn parse_input(
    path: &str,
) -> (
    Vec<Vec<char>>,
    (Coord, Coord),
    Vec<(Coord, Coord)>,
    (Coord, Coord),
) {
    let map: Vec<Vec<char>> = fs::read_to_string(path)
        .expect("File not found")
        .lines()
        .map(|l| l.chars().collect())
        .collect();
    let mut end = (0, 0);
    let mut s = (0, 0);
    let mut strt = vec![(0, 0)];

    for (i, x) in map.iter().enumerate() {
        for (j, &y) in x.iter().enumerate() {
            match y {
                'E' => end = (i as Coord, j as Coord),
                'S' => s = (i as Coord, j as Coord),
                'a' => {
                    // if we're surrounded by other 'a', 'S' or something unreachable, don't even consider this start
                    // they can't even be local optima
                    if !(i > 0
                        && height(&map, ((i - 1) as Coord, j as Coord)) != 1
                        && i < map.len() - 1
                        && height(&map, ((i + 1) as Coord, j as Coord)) != 1
                        && j > 0
                        && height(&map, (i as Coord, (j - 1) as Coord)) != 1
                        && j < map[i].len() - 1
                        && height(&map, (i as Coord, (j + 1) as Coord)) != 1)
                    {
                        strt.push((i as Coord, j as Coord))
                    }
                }
                _ => {}
            };
        }
    }
    (map, s, strt, end)
}

fn part1(map: &[Vec<char>], start: (Coord, Coord), end: (Coord, Coord)) -> usize {
    let mut visited: HashSet<(Coord, Coord)> = HashSet::new();
    let mut distances: HashMap<(Coord, Coord), Height> = HashMap::new();
    distances.insert(start, 0);
    // counting start as first step - will have to subtract later on
    let mut fringe: VecDeque<(Coord, Coord)> = VecDeque::from([start]);

    bfs(map, &mut visited, &mut distances, &mut fringe);

    let height: &Height = distances.get(&end).or(Some(&Height::MAX)).unwrap();
    *height as usize
}

fn height(map: &[Vec<char>], idx: (Coord, Coord)) -> usize {
    let c: char = map[idx.0 as usize][idx.1 as usize];

    match c {
        'S' => 0,
        'E' => 26,
        _ => (c as Coord - b'a') as usize,
    }
}

fn bfs(
    map: &[Vec<char>],
    visited: &mut HashSet<(Coord, Coord)>,
    distances: &mut HashMap<(Coord, Coord), Height>,
    fringe: &mut VecDeque<(Coord, Coord)>,
) {
    while !fringe.is_empty() {
        let curr = fringe.pop_front().unwrap();
        if visited.contains(&curr) {
            continue;
        }
        visited.insert(curr);

        let dist = (*distances.get(&curr).unwrap()) + 1;
        let max_height = height(map, curr) + 1;

        // process all neighbours
        if curr.1 > 0 {
            let left = (curr.0, curr.1 - 1);
            if height(map, left) <= max_height && !visited.contains(&left) {
                fringe.push_back(left);
                distances.insert(left, dist);
            }
        }

        if (curr.1 as usize) < map[curr.0 as usize].len() - 1 {
            let right = (curr.0, curr.1 + 1);
            if height(map, right) <= max_height && !visited.contains(&right) {
                fringe.push_back(right);
                distances.insert(right, dist);
            }
        }

        if curr.0 > 0 {
            let up = (curr.0 - 1, curr.1);
            if height(map, up) <= max_height && !visited.contains(&up) {
                fringe.push_back(up);
                distances.insert(up, dist);
            }
        }

        if (curr.0 as usize) < map.len() - 1 {
            let down = (curr.0 + 1, curr.1);
            if height(map, down) <= max_height && !visited.contains(&down) {
                fringe.push_back(down);
                distances.insert(down, dist);
            }
        }
    }
}

fn part2(map: &[Vec<char>], strt: &[(Coord, Coord)], end: (Coord, Coord), best: usize) -> usize {
    min(
        strt.iter()
            .map(|&start| part1(map, start, end))
            .min()
            .unwrap(),
        best,
    )
}

File: 12/12.py
from collections import deque
from typing import TypeAlias

from timing_util import print_elapsed, timestamp_nano

Coord: TypeAlias = tuple[float, float]


def get_data(content: str) -> tuple[list[list[str]], Coord, list[Coord], Coord]:
    grid = [list(line) for line in content.splitlines()]

    s, end = None, None
    strt = []
    for i, row in enumerate(grid):
        for j, letter in enumerate(row):
            match letter:
                case 'E':
                    end = i, j
                case 'S':
                    s = i, j
                case 'a':
                    if not (
                        i > 0 and height(grid, (i - 1, j)) != 1 and
                        i < len(grid) - 1 and height(grid, (i + 1, j)) != 1 and
                        j > 0 and height(grid, (i, j - 1)) and
                        j > len(grid[i]) - 1 and height(grid, (i, j + 1)) != 1
                    ):
                        strt.append((i, j))

    return grid, s, strt, end


def height(grid: list[list[str]], idx: Coord) -> int:
    match grid[idx[0]][idx[1]]:
        case 'S':
            return 0
        case 'E':
            return 26
        case letter:
            return ord(letter) - ord('a')


def bfs(
    grid: list[list[str]],
    visited: set[Coord],
    distances: dict[Coord, int],
    fringe: deque[Coord]
):
    while fringe:
        curr = fringe.popleft()
        if curr in visited:
            continue
        
        visited.add(curr)
        dist = distances[curr] + 1

        max_height = height(grid, curr) + 1
        curr_x, curr_y = curr
        if curr_y > 0:
            left = curr_x, curr_y - 1
            if height(grid, left) <= max_height and left not in visited:
                fringe.append(left)
                distances[left] = dist
        
        if curr_y < len(grid[curr_x]) - 1:
            right = curr_x, curr_y + 1
            if height(grid, right) <= max_height and right not in visited:
                fringe.append(right)
                distances[right] = dist
        
        if curr_x > 0:
            up = curr_x - 1, curr_y
            if height(grid, up) <= max_height and up not in visited:
                fringe.append(up)
                distances[up] = dist
        
        if curr_x < len(grid) - 1:
            down = curr_x + 1, curr_y
            if height(grid, down) <= max_height and down not in visited:
                fringe.append(down)
                distances[down] = dist


def part1(grid: list[list[str]], start: Coord, end: Coord) -> int | None:
    visited = set()
    distances = {start: 0}
    fringe = deque([start])

    bfs(grid, visited, distances, fringe)

    return distances.get(end, None)


def part2(grid: list[list[str]], strt: list[Coord], end: Coord, best) -> int:
    return min(best, min(filter(bool, (part1(grid, start, end) for start in strt))))


if __name__ == '__main__':
    start = timestamp_nano()

    with open('12/input.txt') as in_file:
        grid, s, strt, end = get_data(in_file.read())

    p1 = part1(grid, s, end)
    print(f'part1: {p1}')
    print(f'part2: {part2(grid, strt, end, p1)}')

    print_elapsed(start)
